*In this WIP tutorial, we will investigate how several gift and candy stores on Oxford Street are connected.*

Recent investigations have highlighted lots of delinquient shops on Oxford Street selling candy, luggage and gifts. Whilst some of these businesses can be identified within Companies House, this task can be challenging due to the company names and addresses often not corresponding to the name and address of the physical shops. Additionally many of the owners of these businesses file incorrect or missing information. However, we can still investigate these companies to explore the ways in which they may be connected. 

In a [recent article]('https://www.ft.com/candy) by the Financial Times, several candy shops were mapped and discussed. I have done my best to find the records for these companies in Companies House:

- Kingdom of Sweets (270 Oxford Street): 11004735
- Western Crown Limited (470-482 Oxford Sreet): 13455377
- City Souvenirs aka Candy Surprise (537, 539 Oxford Street): 08658716
- Crest of London (443 Oxford Street): 13760753
- Surprise Surprise (271 Oxford Street): 11688369
- City Luggage Ltd (273 Oxford Street): 07888549
- Souvenirs of London (137 Oxford Street): 03114319 

Lets now see if any of them are connected...

In [1]:
candy_shops = ['11004735', '13455377', '08658716', '13760753', '11688369', '07888549', '03114319']

In [2]:
from crawler import *

In [3]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 150)
from IPython.display import clear_output

Lets investigate 'WESTERN CROWN LIMITED'.

In [4]:
network = Ownership_Network(company_id="13455377")

We're now going to build a network of companies, people and addresses starting from Western Crown Limited:

In [5]:
network.company_ids

company_id  n    edge  node node_type
0   13455377  0  Origin  None   Company

The following 'hop' method will take the company_id and find addresses and people connected to it. Lets begin by performing a single network hop:

In [6]:
network.hop(1)

Processed 1/1 companies


Now if we check out the `officer_ids` and `addresses` DataFrames in our class, we can check to see if some new entries have been added:

In [7]:
network.addresses

address  n  \
0  470-482 Oxford St Oxford Street London W1C 1LA...  1   
1  407-409 Oxford St Artex Computers Limited Oppo...  1   
2  474 Oxford St Oxford Street London W1C 1LA Eng...  1   
3  Flat 7 594 Commercial Road London E14 7JR England  1   

                            edge      node node_type  
0  Officer Corresponance Address  13455377   Company  
1     Company Historical Address  13455377   Company  
2     Company Historical Address  13455377   Company  
3     Company Historical Address  13455377   Company

In [8]:
network.officer_ids

officer_id  n     edge                 name      node  \
0  FHa6xpWl842N03mouAYtCciAt_8  1  Officer  Isfahan CHOMBO KADE  13455377   

  node_type  
0   Company

It looks like some additional addresses appeared. If we check out the `edge` column, we can see the description for how these enteries were added is because these addresses used to be where company 13455377 aka West Crown Limited was registered. We can see that no new officers were added, which is due to the fact that the company has only ever had one officer. 

Lets perform another hop, which will take all the enteries with n=1 and search for people, companies and addresses connected to them:

In [9]:
network.hop(1)

Processed 1/1 officers


We can now see some more addresses coming into view tied to the psc and officer for West Crown Limited:

In [10]:
network.addresses

address  n  \
0  470-482 Oxford St Oxford Street London W1C 1LA...  1   
1  407-409 Oxford St Artex Computers Limited Oppo...  1   
2  474 Oxford St Oxford Street London W1C 1LA Eng...  1   
3  Flat 7 594 Commercial Road London E14 7JR England  1   
4  Flat 47 594 Commercial Road London E14 7JR Eng...  2   
5                      Flat 7 London E14 7JR England  2   
6  Flat 142-144 Oxford Street London W1D 1LZ England  2   
7  25 Fabian Bell Tower Pancras Way London E3 2SD...  2   

                            edge                         node node_type  
0  Officer Corresponance Address                     13455377   Company  
1     Company Historical Address                     13455377   Company  
2     Company Historical Address                     13455377   Company  
3     Company Historical Address                     13455377   Company  
4            Appointment Address  FHa6xpWl842N03mouAYtCciAt_8    Person  
5            Appointment Address  FHa6xpWl842N03mouAYtCciAt_8    Person  
6            Appointment Address  FHa6xpWl842N03mouAYtCciAt_8    Person  
7            Appointment Address  FHa6xpWl842N03mouAYtCciAt_8    Person

Interestingly it looks like there's multiple instances of this officer's name with the same birthday registering businesses in companies house. We log these as 'Duplicate Officer':

In [11]:
network.officer_ids

officer_id  n                edge  \
0  FHa6xpWl842N03mouAYtCciAt_8  1             Officer   
1  k-TdhEwo_aWTkyZviosmuFYu00o  2  Officer at Address   
2  Jpod88ZD25UttBRLoQaoTeA4KC4  2   Duplicate Officer   
3  flpImMENIgEVKD6l_uuerTgiY10  2   Duplicate Officer   
4  WWGUAanL13te4Yc53Oxyh-HNes8  2   Duplicate Officer   

                        name  \
0        Isfahan CHOMBO KADE   
1  Brent Christopher LINDSAY   
2        Isfahan CHOMBO KADE   
3        Isfahan CHOMBO KADE   
4        Isfahan CHOMBO KADE   

                                                node node_type  
0                                           13455377   Company  
1  Flat 7 594 Commercial Road London E14 7JR England   Address  
2                        FHa6xpWl842N03mouAYtCciAt_8    Person  
3                        FHa6xpWl842N03mouAYtCciAt_8    Person  
4                        FHa6xpWl842N03mouAYtCciAt_8    Person

There's also some more companies. If we want to also get the names and additional data for these companies, we can use the `get_company_from_id` method. We can get a single companies info or pass the entire `company_ids` DataFrame by not providing any input parameters. Company info is then stored seperately in the `companies` DataFrame as shown below. We can already see a number of other gift themed shops appearing in the results!

In [12]:
network.company_ids

company_id  n             edge  \
0   13455377  0           Origin   
1   07478039  2  Company Address   
2   12126349  2  Company Address   
3   10436966  2  Company Address   
4   11461249  2      Appointment   
5   13376158  2      Appointment   
6   12005370  2      Appointment   

                                                node node_type  
0                                               None   Company  
1  Flat 7 594 Commercial Road London E14 7JR England   Address  
2  Flat 7 594 Commercial Road London E14 7JR England   Address  
3  Flat 7 594 Commercial Road London E14 7JR England   Address  
4                        FHa6xpWl842N03mouAYtCciAt_8    Person  
5                        FHa6xpWl842N03mouAYtCciAt_8    Person  
6                        FHa6xpWl842N03mouAYtCciAt_8    Person

In [13]:
network.get_company_from_id()
network.companies

company_number                              n  \
0       13455377                              0   
1       07478039  1    2
Name: n, dtype: object   
2       12126349  2    2
Name: n, dtype: object   
3       10436966  3    2
Name: n, dtype: object   
4       11461249  4    2
Name: n, dtype: object   
5       13376158  5    2
Name: n, dtype: object   
6       12005370  6    2
Name: n, dtype: object   

                          company_name company_status date_of_creation  \
0                WESTERN CROWN LIMITED         active       2021-06-14   
1              GIFTS OF LONDON LIMITED         active       2010-12-23   
2              SQUIRREL RETAIL LIMITED      dissolved       2019-07-29   
3  LITTLE RED SQUIRREL FASHION LIMITED         active       2016-10-19   
4                 AFFIRM GIFTS LIMITED      dissolved       2018-07-12   
5                 GRAND EMPIRE LIMITED         active       2021-05-05   
6                       CANDY CANE LTD      dissolved       2019-05-20   

                                       etag has_charges  \
0  62747592dd17796cf54645056187d05065962c5e       False   
1  19e95cd62708edc20d50dcdddec5696bba1ab265       False   
2  81c34a615d3edcb5f665747f7675ebdaf5b1861c       False   
3  30d6bee339fc7857d19673edb1c8063a702671bc       False   
4  127d334c4644364f9982f509497ce7eabad97d51       False   
5  01cb599b001d570b521aa5ea228c50c8ae65281f       False   
6  cde3d72b19642a55f0ce9b8351020dc5a475ddfd       False   

  has_insolvency_history has_super_secure_pscs   jurisdiction  \
0                  False                 False  england-wales   
1                  False                 False  england-wales   
2                  False                 False  england-wales   
3                  False                 False  england-wales   
4                  False                 False  england-wales   
5                  False                 False  england-wales   
6                  False                 False  england-wales   

  registered_office_is_in_dispute sic_codes type  \
0                           False   [47190]  ltd   
1                           False   [47190]  ltd   
2                           False   [47190]  ltd   
3                           False   [47190]  ltd   
4                           False   [47190]  ltd   
5                           False   [47190]  ltd   
6                           False   [47240]  ltd   

  undeliverable_registered_office_address can_file  \
0                                   False     True   
1                                   False     True   
2                                   False    False   
3                                   False     True   
4                                   False    False   
5                                   False     True   
6                                   False    False   

                                             edge  \
0                                          Origin   
1  1    Company Address
Name: edge, dtype: object   
2  2    Company Address
Name: edge, dtype: object   
3  3    Company Address
Name: edge, dtype: object   
4      4    Appointment
Name: edge, dtype: object   
5      5    Appointment
Name: edge, dtype: object   
6      6    Appointment
Name: edge, dtype: object   

               links.filing_history         links.self  \
0  /company/13455377/filing-history  /company/13455377   
1  /company/07478039/filing-history  /company/07478039   
2  /company/12126349/filing-history  /company/12126349   
3  /company/10436966/filing-history  /company/10436966   
4  /company/11461249/filing-history  /company/11461249   
5  /company/13376158/filing-history  /company/13376158   
6  /company/12005370/filing-history  /company/12005370   

               links.officers  \
0  /company/13455377/officers   
1  /company/07478039/officers   
2  /company/12126349/officers   
3  /company/10436966/officers   
4  /company/11461249/officers   
5  /company/13376158/officers   
6  /company/12005370/officers   

 

We can check how many hops we've performed by querying the `n` attribute:

In [14]:
network.n

2

Lets now perform 3 hops. It will inevitably take a bit longer this time, you might want to make a cup of tea. Last time I checked it completed in around 15 mins: 

In [15]:
network.hop(3)

Processed 74/75 officers


Additionally we can also fill in all the company info before we take a deeper look. This will also probably take around 8 mins:

In [26]:
network.get_company_from_id()

After 5 hops we've now processed thousands of companies, people and addresses. Lets now see if the results contain any candy shops at the addresses mentioned in the FT:

In [34]:
network.companies[network.companies['registered_office_address.address_line_1'].str.contains("Oxford Street")]

company_number                                n  \
0         13455377                                0   
5         13376158    5    2
Name: n, dtype: object   
7         12672841    7    3
Name: n, dtype: object   
9         13640427    9    3
Name: n, dtype: object   
12        14210992   12    3
Name: n, dtype: object   
185       12415826  185    4
Name: n, dtype: object   
186       10734212  186    4
Name: n, dtype: object   
187       11843228  187    4
Name: n, dtype: object   
188       13865359  188    4
Name: n, dtype: object   
189       11864536  189    4
Name: n, dtype: object   
191       11031538  191    4
Name: n, dtype: object   
201       13640435  201    4
Name: n, dtype: object   
205       13287820  205    4
Name: n, dtype: object   
206       12387042  206    4
Name: n, dtype: object   
207       11647876  207    4
Name: n, dtype: object   
359       10116914  359    5
Name: n, dtype: object   
360       08474569  360    5
Name: n, dtype: object   
361       08576317  361    5
Name: n, dtype: object   
362       09071473  362    5
Name: n, dtype: object   
363       08632458  363    5
Name: n, dtype: object   
364       10087753  364    5
Name: n, dtype: object   
365       11593230  365    5
Name: n, dtype: object   
366       07548588  366    5
Name: n, dtype: object   
367       08282993  367    5
Name: n, dtype: object   
368       13964334  368    5
Name: n, dtype: object   
369       09373180  369    5
Name: n, dtype: object   
370       09675396  370    5
Name: n, dtype: object   
371       07452617  371    5
Name: n, dtype: object   
374       13967497  374    5
Name: n, dtype: object   
375       10231441  375    5
Name: n, dtype: object   
376       14472151  376    5
Name: n, dtype: object   
377       08645355  377    5
Name: n, dtype: object   
379       13795800  379    5
Name: n, dtype: object   
380       13857378  380    5
Name: n, dtype: object   
381       11468719  381    5
Name: n, dtype: object   
601       10941293  601    5
Name: n, dtype: object   
602       11647421  602    5
Name: n, dtype: object   
629       08949849  629    5
Name: n, dtype: object   
633       07570906  633    5
Name: n, dtype: object   
661       11588633  661    5
Name: n, dtype: object   
663       11869360  663    5
Name: n, dtype: object   
665       10540083  665    5
Name: n, dtype: object   
666       11016980  666    5
Name: n, dtype: object   
667       08795415  667    5
Name: n, dtype: object   
670       05340691  670    5
Name: n, dtype: object   
695       10777885  695    5
Name: n, dtype: object   
696       11439227  696    5
Name: n, dtype: object   
800       12268339  800    5
Name: n, dtype: object   

                             company_name company_status date_of_creation  \
0                   WESTERN CROWN LIMITED         active       2021-06-14   
5                    GRAND EMPIRE LIMITED         active       2021-05-05   
7                     LOTUS WHITE LIMITED      dissolved       2020-06-15   
9                       MY APARTMENTS LTD         active       2021-09-23   
12                         FANCYGREEN LTD         active       2022-07-04   
185                       CANDYSTREET LTD         active       2020-01-21   
186              XEE ASSET MANAGEMENT LTD      dissolved       2017-04-21   
187                       GRAND STORE LTD      dissolved       2019-02-23   
188                       SPARK LABEL LTD         active       2022-01-22   
189                       LUMS SWEETS LTD         active       2019-03-06   
191                      COOL MIX LIMITED      dissolved       2017-10-25   
201              MY LONDON APARTMENTS LTD         active       2021-09-23   
205                    MOODY MOON LIMITED         active       2021-03-24   
206                    GIFT PUNCH LIMITED      dissolved       2020-01-06   
207                     CEEKEY LONDON LTD         active       2018-10-29   
359        LONDON HOT ACCESSORIES LIMITED      dissolved       2016-04-11   

We can see many more gift and candy shops registered to Oxford Street addresses appearing:
    
- GIFT 4 YOU LIMITED (11439227): 4 Oxford Street 
- GIFTS OF GLORY LTD (12268339): 146/148 Oxford St Oxford Street
- CITY GIFTS LIMITED (05340691): 158 Oxford Street
- GLOBAL GIFTS 4 U LTD (08795415): 158 Oxford Street
- ROCK GIFTS LTD (11588633): 524 Oxford Street
- PLANET SOUVENIRS (UK) LIMITED (07570906): 451 Oxford Street
- WEST ONE SOUVENIRS LIMITED (08949849): 161-163 Oxford Street
- VENGAT GIFT LTD (11647421): 263 265 Oxford Street
- ASUSGIFTS LIMITED	 (13857378): 37-39 Oxford Street
- BRITCO GIFTS LIMITED (14472151): 37-39 Oxford Street
- FANCYGREEN LTD (14210992): 37-39 Oxford Street
- CANDYSTREET LTD (12415826	): 146-148 Oxford Street
- LUMS SWEETS LTD (11864536): 146-148 Oxford Street	
- GIFT PUNCH LIMITED (12387042): 37-39 Oxford Street
- LONDON HOT ACCESSORIES LIMITED (10116914): 537 Oxford Street
- GIFTNET LTD (11593230): 37 -39 Oxford Street

If we want to see how any of these businesses is connected to the original business (West Crown Limited), we can use the `find_path` method passing the company_id of the company we wish to find the connection for:

In [37]:
network.find_path("11439227")

WESTERN CROWN LIMITED -> Isfahan CHOMBO KADE (Officer) -> Isfahan CHOMBO KADE (Duplicate Officer) -> BEST OF LONDON LIMITED (Appointment) -> Zahidullah HUSSAIN KHIL (Officer) -> GIFT 4 YOU LIMITED (Appointment) 


In [38]:
network.find_path("10116914")

WESTERN CROWN LIMITED -> Isfahan CHOMBO KADE (Officer) -> Isfahan CHOMBO KADE (Duplicate Officer) -> BEST OF LONDON LIMITED (Appointment) -> 537 Oxford Street London W1C 2QP United Kingdom (Company Historical Address) -> LONDON HOT ACCESSORIES LIMITED (Company Address) 
